public/js/test/mochitest/head.js

```diff
@@ -0,0 +1,176 @@
+    while (count < eventRepeat) {
+      yield _afterDispatchDone(dbg.store, type);
+      count++;
+      info(type + " dispatched " + count + " time(s)");
+    }
+  });
+}
+
+function waitForThreadEvents(dbg, eventName) {
+  info("Waiting for thread event '" + eventName + "' to fire.");
+  const thread = dbg.toolbox.threadClient;
+
+  return new Promise(function(resolve, reject) {
+    thread.addListener(eventName, function onEvent(eventName, ...args) {
+      info("Thread event '" + eventName + "' fired.");
+      thread.removeListener(eventName, onEvent);
+      resolve.apply(resolve, args);
+    });
+  });
+}
+
+function waitForState(dbg, predicate) {
+  return new Promise(resolve => {
+    const unsubscribe = dbg.store.subscribe(() => {
+      if (predicate(dbg.store.getState())) {
+        unsubscribe();
+        resolve();
+      }
+    });
+  });
+}
+
+const waitForPaused = Task.async(function* (dbg) {
+  // We want to make sure that we get both a real paused event and
+  // that the state is fully populated. The client may do some more
+  // work (call other client methods) before populating the state.
+  return Promise.all([
+    yield waitForThreadEvents(dbg, "paused"),
+    yield waitForState(dbg, state => dbg.selectors.getPause(state))
+  ]);
+});
+
+const initDebugger = Task.async(function* (url, ...sources) {
+  const toolbox = yield openNewTabAndToolbox(url, "jsdebugger");
+  const win = toolbox.getPanel("jsdebugger").panelWin;
+  const store = win.Debugger.store;
+  const { getSources } = win.Debugger.selectors;
+
+  const dbg = {
+    actions: win.Debugger.actions,
+    selectors: win.Debugger.selectors,
+    getState: store.getState,
+    store: store,
+    toolbox: toolbox,
+    win: win
+  };
+
+  yield Promise.all(sources.map(url => {
+    return waitForState(dbg, state => {
+      return getSources(state).some(s => s.get("url").includes(url));
+    });
+  }));
+
+  return dbg;
+});
+
+function findSource(dbg, url) {
+  const sources = dbg.selectors.getSources(dbg.getState());
+  const source = sources.find(s => s.get("url").includes(url));
+
+  if(!source) {
+    throw new Error("Unable to find source: " + matchingStr);
+  }
+
+  return source.toJS();
+}
+
+function selectSource(dbg, url) {
+  info("Selecting source: " + url);
+  const source = findSource(dbg, url);
+  dbg.actions.selectSource(source.id);
+
+  return waitForDispatch(dbg, "LOAD_SOURCE_TEXT");
+}
+
+function stepOver(dbg) {
+  info("Stepping over");
+  dbg.actions.stepOver();
+  return waitForPaused(dbg);
+}
+
+function stepIn(dbg) {
+  info("Stepping in");
+  dbg.actions.stepIn();
+  return waitForPaused(dbg);
+}
+
+function stepOut(dbg) {
+  info("Stepping out");
+  dbg.actions.stepOut();
+  return waitForPaused(dbg);
+}
+
+function resume(dbg) {
+  info("Resuming");
+  dbg.actions.resume();
+  return waitForThreadEvents(dbg, "resumed");
+}
+
+function addBreakpoint(dbg, sourceId, line, col) {
+  return dbg.actions.addBreakpoint({ sourceId, line, col });
```

Comment:
(Wanted to reply longer here and had to step out)

> I think it is good to keep the API similar to what the components call, that way we exercise the same behavior. We can't guarantee it's the same, but it's probably good to try.

Those actions are what they call :) But we can't guarantee that we are going to call the actions with exactly the same arguments, but I think that's ok. Our actions should be usable like this. They should just take simple arguments and work.

The fact that the `addBreakpoint` action requires that extra function is a bad thing (I know I implemented it). We need to fix that. But for the majority of the tests, it really doesn't matter that the text snippet isn't there. I'd like to write it this way, and just try improve the `addBreakpoint` so it has the text correctly without requiring this extra work. For now, we'll just make sure we add the line snippet when testing the breakpoint pane. It's a small detail that doesn't really matter, and it's just a bad API that we need to fix.

> The addBreakpoint command I setup assumes it's the selected source and passes along the getTextForLine param for that reason.

The selected source thing is a bigger question. I completely understand where you are coming from, but let me try to explain why I don't think we should do it.

We don't drive the UI by actually clicking on the DOM in every test for a reason. For any given test, they are really trying to make sure a specific workflow is correct (and ignore all the other details). Redux actions are a super nice way to do this, even if it's not literally the same workflow that the UI goes through. As long as we make sure adding a breakpoint on the current selected source works, it shouldn't matter if adding a breakpoint is on the selected source or not in other tests. If we really wanted to test the UI as the user uses it every time, we'd be clicking DOM elements.

I actually really like the idea of being able to set breakpoints on other sources than the selected one, and I want to do that some day. Something like a "command bar" that lets you say "break <url> <line>". That would be awesome. There's no reason we should tie together a simple action like that with changing out the entire editor, loading source texts, etc.

In the old debugger, there are many tests that set breakpoints in several sources, and then continue through and make sure all of them are hit. These tests are a lot simpler if you can just set a breakpoint on any source. It makes it easier to test complex scenarios like "set a breakpoint on source X and make sure when the breakpoint it hit that it opens and selects that source".

So that's my vision of the tests: generally the commands map straight to redux actions, but with a few niceties like the stepping ones waiting for the right `paused` event. This lets tests set up all kinds of workflows, whatever they want, even if many of them aren't real user workflows. We can make sure that things work in weird ways without the commands getting in the way of a specific workflow.

Just wanted to explain my view of things; I totally get where you're coming from. The good thing is this is something that we could pretty easily change later just by changing the `addBreakpoint` command if we are missing bugs for some reason.